File: inn/inn_hotels/web_form/room_booking/room_booking.js
frappe.ready(function() {
	var date_end;
	var date_start;
	var today = new Date();
	today.setHours(2);

	frappe.web_form.on("start", (field, value) => {
		date_start = new Date(value);
		if (date_start < today){
			field.set_value(0)
			frappe.msgprint("Start date cannot before today")
		} else if (date_end == undefined || date_start == undefined){
			return
		} else if (date_end <= date_start){
			field.set_value(0)
			frappe.msgprint("Start date cannot after end date")
		} else {
			get_available_room_and_rate()
		}
	});

	frappe.web_form.on("end", (field, value) => {
		date_end = new Date(value);
		date_end.setHours(1)
		if (date_end < today){
			field.set_value(0)
			frappe.msgprint("End date cannot before tommorow")
		} else if (date_start == undefined || date_end == undefined){
			return
		} else if (date_end <= date_start){
			field.set_value(0)
			frappe.msgprint("End date must after start date")
		} else {
			get_available_room_and_rate()
		}
		
	});
	
	frappe.web_form.on("number_of_rooms", (field, value) => {
		if (value < 1) {
			field.set_value(1)
			frappe.msgprint("Number of room must greater than 0")
		} else if (date_end == undefined || date_start == undefined){
			return
		} else {
			get_available_room_and_rate()
		}
	})

	frappe.web_form.on("room_type_custom", (field, value) => {
		frappe.web_form.set_value(["price"], options_value[value] * frappe.web_form.get_value("number_of_rooms"))
	})
})

var options_value = {}

function get_available_room_and_rate() {
	frappe.web_form.set_value("price", undefined)
	frappe.web_form.set_value("room_type_custom", undefined)
	frappe.call({
		method: "inn.inn_hotels.web_form.room_booking.room_booking.get_available_room_and_rate",
		args: {
			start_date: frappe.web_form.get_value(["start"]),
			end_date: frappe.web_form.get_value(["end"]),
			num_room: frappe.web_form.get_value(["number_of_rooms"])
		},
		callback: (r) => {
			var option = []
			for (ii of r.message){
				room_type = "type: " + ii.room_type
				bed_type = "bed: " + ii.bed_type
				allow_smoke = ii.allow_smoke == "Yes" ? "smoking room" : "non-smoking room"
				incl_breakfast =  ii.incl_breakfast ? "breakfast" :  "non-breakfast"
				price = ii.price.toLocaleString()
				key = room_type + ", " + bed_type + ", " + allow_smoke + ", " + incl_breakfast + " --- Rp. "+ price

				options_value[key] = ii.price  
				option.push(key)
			}
			frappe.web_form.set_df_property("room_type_custom", "options", option)
		}
	})
}

File: inn/inn_hotels/web_form/room_booking/room_booking.py
from datetime import timedelta, date, datetime
from inn.helper import daterange
import operator
import frappe


def get_context(context):
	# do your magic here
	pass

class InnRoomBookingChoice:
	__slots__ = "room_type", "bed_type", "allow_smoke", "incl_breakfast", "price"

	def __init__(self, row):
		self.room_type = row[0]
		self.bed_type = row[1]
		self.allow_smoke = row[2]

	def add_rate(self, prices):
		self.price = float(prices["final_total_rate_amount"])
		self.incl_breakfast = prices["final_breakfast_rate_amount"] > 0

	def adjust_price_with_day_and_room(self, num_of_room: int, num_of_night: int):
		self.price = self.price * num_of_night * num_of_room

	def toJSON(self):
		return {key : getattr(self, key, None) for key in self.__slots__}
	
	def __getitem__(self, key):
		return getattr(self, key)
				 


def tup_key_gen(row):
	return (row[1], row[2], row[3])


def get_rate(available_room, num_of_room: int, num_of_night: int) -> list:

	default_group_guest = "Guest Booking Group"
	setting_group_guest = frappe.db.get_value(doctype="Inn Hotels Setting", fieldname="guest_booking_group")
	if setting_group_guest != None:
		default_group_guest = setting_group_guest

	result = []

	# prevent multiple query with same filter
	room_types = set()
	for ii in available_room:
		room_types.add(ii[0])
		

	for room_type in room_types:
		prices = frappe.db.get_list(ignore_permissions=True, doctype="Inn Room Rate", filters={"customer_group": default_group_guest, "room_type": room_type}, fields=["final_total_rate_amount", "final_breakfast_rate_amount"])
		if len(prices) == 0:
			continue

		for price in prices:
			for jj in available_room:
				if jj[0] == room_type:
					elem = InnRoomBookingChoice(jj)
					elem.add_rate(price)
					elem.adjust_price_with_day_and_room(num_of_room, num_of_night)
					result.append(elem)

	return sorted(result, key = operator.itemgetter("room_type", "bed_type", "allow_smoke"))
	
def convert_json(obj):
	return [x.toJSON() for x in obj]


@frappe.whitelist(allow_guest=True)
def get_image_carousel():
	file = frappe.db.get_list("File", filters={"Folder":"Home/Carousel"}, fields=["file_url"], ignore_permissions=True)
	return file

@frappe.whitelist(allow_guest=True)
def get_available_room_and_rate(start_date, end_date, num_room):
	if type(num_room) == str:
		num_room = int(num_room)

	# get number of room with same room type and bed type
	default_availability = frappe.db.sql(
		'SELECT count(*), room_type, bed_type, allow_smoke '
		'from `tabInn Room` group by room_type, bed_type, allow_smoke'
		)
	available_room = {}
	for ii in default_availability:
		room_key = tup_key_gen(ii)
		available_room[room_key] = ii[0]

	start_date = datetime.strptime(start_date, "%Y-%m-%d")
	end_date = datetime.strptime(end_date, "%Y-%m-%d")
	num_night = end_date - start_date
	for curr_date in daterange(start_date, end_date):
		used_availability = frappe.db.sql(
			"select count(*), room_type, bed_type, allow_smoke from `tabInn Room` as ir "
			"join `tabInn Room Booking` as irb on irb.room_id = ir.name where status not in ('Finished', 'Canceled') "
			"and irb.start <= %(date)s and irb.end > %(date)s group by bed_type, room_type, allow_smoke"
		, values={"date":curr_date}, as_dict=0)

		# reduce room number because being used in this date
		unusable_room = {}
		for ii in used_availability:
			room_key = tup_key_gen(ii)
			unusable_room[room_key] = ii[0]

		# check if jumlah tipe kamar yang tersedia memenuhi jumlah kamar yang diminta
		not_enough_quantity = []
		for ii in available_room:
			kamar_sisa = available_room[ii]
			if ii in unusable_room:
				kamar_sisa -= unusable_room[ii]
			if kamar_sisa < int(num_room):
				not_enough_quantity.append(ii)

		for ii in not_enough_quantity:
			available_room.pop(ii)

	result = get_rate(available_room, num_room, num_night.days)
	result = convert_json(result)

	return result

